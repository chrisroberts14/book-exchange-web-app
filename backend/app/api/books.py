"""Module for book endpoints."""

from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from starlette.status import HTTP_201_CREATED, HTTP_404_NOT_FOUND, HTTP_204_NO_CONTENT

from backend.app.api_models import BookOut, BookIn, BookPatch
from backend.app.core.db import get_db
from backend.app.db_models import BookDb

books = APIRouter()


@books.post("/", status_code=HTTP_201_CREATED)
async def create_book(book: BookIn, db: Session = Depends(get_db)) -> BookOut:
    """
    Create a book.

    :param db:
    :param book: book to create
    :return:
    """
    db_book = BookDb(**book.dict())
    return BookDb.create(db, db_book)


@books.get("/")
async def get_all_books(db: Session = Depends(get_db)) -> list[BookOut]:
    """
    Get all books.

    :return:
    """
    return BookDb.get_all(db)


@books.get("/{book_id}")
async def get_book(book_id: UUID, db: Session = Depends(get_db)) -> BookOut:
    """
    Get a book by id.

    :param book_id: id of the book
    :param db: database session
    :return:
    """
    book = BookDb.get_by_id(db, book_id)
    if book is None:
        raise HTTPException(status_code=HTTP_404_NOT_FOUND, detail="Book not found")
    return book


@books.patch("/{book_id}")
async def update_book(
    book_id: UUID, book_patch: BookPatch, db: Session = Depends(get_db)
) -> BookOut:
    """
    Update a book.

    :param book_id:
    :param book_patch:
    :param db:
    :return:
    """
    book = BookDb.get_by_id(db, book_id)
    if book is None:
        raise HTTPException(status_code=HTTP_404_NOT_FOUND, detail="Book not found")
    return BookDb.update(db, book_patch, book_id)


@books.delete("/{book_id}", status_code=HTTP_204_NO_CONTENT)
async def delete_book(book_id: UUID, db: Session = Depends(get_db)) -> None:
    """
    Delete a book.

    :param book_id:
    :param db:
    :return:
    """
    book = BookDb.get_by_id(db, book_id)
    if book is None:
        raise HTTPException(status_code=HTTP_404_NOT_FOUND, detail="Book not found")
    BookDb.delete(db, book_id)
