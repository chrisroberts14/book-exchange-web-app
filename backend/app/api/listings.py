"""Module for listings endpoints."""

from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from starlette.status import HTTP_201_CREATED, HTTP_404_NOT_FOUND

from backend.app.api_models import ListingIn, ListingOut, ListingPatch
from backend.app.core.db import get_db
from backend.app.db_models import ListingDb

listings = APIRouter()


@listings.get("/", response_model=list[ListingOut])
async def get_listings(db: Session = Depends(get_db)) -> list[ListingOut]:
    """
    Get all listings.

    :return:
    """
    return ListingDb.get_all(db)


@listings.post("/", status_code=HTTP_201_CREATED)
async def create_listing(data: ListingIn, db: Session = Depends(get_db)) -> ListingOut:
    """
    Create a listing.

    :return:
    """
    return ListingDb.create(db, ListingDb(**data.dict()))


@listings.get("/{listing_id}", response_model=ListingOut)
async def get_listing(listing_id: UUID, db: Session = Depends(get_db)) -> ListingOut:
    """
    Get a listing by id.

    :param listing_id: id of the listing
    :param db: database session
    :return:
    """
    listing = ListingDb.get_by_id(db, listing_id)
    if listing is None:
        raise HTTPException(status_code=HTTP_404_NOT_FOUND, detail="Listing not found")
    return listing


@listings.patch("/{listing_id}")
async def update_listing(
    listing_id: UUID, data: ListingPatch, db: Session = Depends(get_db)
) -> ListingOut:
    """
    Update a listing.

    :return:
    """
    listing = ListingDb.get_by_id(db, listing_id)
    if listing is None:
        raise HTTPException(status_code=HTTP_404_NOT_FOUND, detail="Listing not found")
    return ListingDb.update(db, data, listing_id)
