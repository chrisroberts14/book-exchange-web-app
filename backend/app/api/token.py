"""Module for logging in and getting a token."""

from typing import Annotated

from fastapi import APIRouter, Depends, HTTPException
from fastapi.security import OAuth2PasswordRequestForm, OAuth2PasswordBearer
from sqlalchemy.orm import Session
from starlette.status import HTTP_400_BAD_REQUEST, HTTP_401_UNAUTHORIZED

from backend.app.api_models import UserOut
from backend.app.common import hash_password
from backend.app.core.db import get_db
from backend.app.db_models import UserDb

token = APIRouter()

oauth2_scheme = OAuth2PasswordBearer(tokenUrl="token/token")


@token.get("/test_auth/")
async def test_auth(tkn: str = Depends(oauth2_scheme)):
    """
    Test endpoint for authentication.

    :param tkn:
    :return:
    """
    return {"token": tkn}


@token.post("/token")
async def login(
    form_data: Annotated[OAuth2PasswordRequestForm, Depends()],
    db: Session = Depends(get_db),
):
    """
    Get a token.

    :param form_data:
    :param db:
    :return:
    """
    user = UserDb.get_user_by_username(db, form_data.username)
    if user is None:
        raise HTTPException(
            status_code=HTTP_400_BAD_REQUEST, detail="Incorrect username or password"
        )
    hashed_password = hash_password(form_data.password)
    if hashed_password != user.hashed_password:
        raise HTTPException(
            status_code=HTTP_400_BAD_REQUEST, detail="Incorrect username or password"
        )
    return {"access_token": user.username, "token_type": "bearer"}


def fake_decode_token(db: Session, tkn: str):
    """
    Decode token stub.

    TODO: Replace with real decode using JWT
    :param db:
    :param tkn:
    :return:
    """
    return UserDb.get_user_by_username(db, tkn)


async def get_current_user(
    tkn: Annotated[str, Depends(oauth2_scheme)], db: Session = Depends(get_db)
):
    """
    Get current user.

    :param tkn:
    :param db:
    :return:
    """
    user = fake_decode_token(db, tkn)
    if not user:
        raise HTTPException(
            status_code=HTTP_401_UNAUTHORIZED,
            detail="Invalid authentication credentials",
            headers={"WWW-Authenticate": "Bearer"},
        )
    return user


@token.get("/me", response_model=UserOut)
async def get_me(
    current_user: Annotated[UserOut, Depends(get_current_user)],
) -> UserOut:
    """
    Get the current user.

    :return:
    """
    return current_user
