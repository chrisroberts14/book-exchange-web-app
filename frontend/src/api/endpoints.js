const address = "http://localhost:8000";

export const loginEndpoint = address + "/auth/login";
export const signupEndpoint = address + "/users/";
export const getAllBooksEndpoint = address + "/books/";
export const getAllListingsEndpoint = address + "/listings/";
