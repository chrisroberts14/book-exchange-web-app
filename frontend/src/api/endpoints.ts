const address = "http://localhost:8000";

export const loginEndpoint: string = address + "/auth/login";
