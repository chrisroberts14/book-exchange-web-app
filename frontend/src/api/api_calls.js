import axios from 'axios';
import {loginEndpoint} from "./endpoints.js";


export const login = async (credentials) => {
    const response = await axios.post(loginEndpoint, new URLSearchParams({
        'username': credentials.username,
        'password': credentials.password
    }));
    return response.data;
}
