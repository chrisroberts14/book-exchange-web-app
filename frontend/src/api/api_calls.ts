import axios from 'axios';
import {loginEndpoint} from "./endpoints";

type LoginCredentials = {
    username: string,
    password: string
}

export const login = async (credentials: LoginCredentials) => {
    const response = await axios.post(loginEndpoint, new URLSearchParams({
        'username': credentials.username,
        'password': credentials.password
    }));
    console.log('Login response:', response.data);
    return response.data;
}
