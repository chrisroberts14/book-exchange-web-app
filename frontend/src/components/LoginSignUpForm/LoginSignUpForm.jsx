import React, { useState } from "react";
import {
  Tabs,
  Tab,
  Input,
  Link,
  Button,
  Card,
  CardBody,
} from "@nextui-org/react";
import{ login, signup } from "../../api/api_calls.js";
import { ErrorMessageBar } from "../ErrorMessageBar/ErrorMessageBar";

export const LoginSignUpForm = ({ setUser, setToken }) => {
  const [formSelected, setFormSelected] = useState("login");
  const [username, setUsername] = useState("");
  const [password, setPassword] = useState("");
  const [email, setEmail] = useState("");
  const [isError, setIsError] = useState(false);
  const [errorMessage, setErrorMessage] = useState("");

  const handleLogin = async (e) => {
    e.preventDefault();
    const response = await login({username: username, password: password});
    if (response.access_token !== undefined) {
      setUser(response.username);
      setUsername("");
      setEmail("");
      setPassword("");
      window.localStorage.setItem("token", response.access_token);
      window.localStorage.setItem("username", response.username);
    } else {
      setIsError(true);
      setTimeout(() => setIsError(false), 5000);
      setErrorMessage(response.errorMessage);
    }
  }

  const handleSignUp = async (e) => {
    e.preventDefault();
    const response = await signup({username: username, password: password, email: email});
      if (response.username !== undefined) {
        setUser(response.username);
        setUsername("");
        setEmail("");
        setPassword("");
      } else {
        setIsError(true);
        setTimeout(() => setIsError(false), 5000);
        setErrorMessage(response.errorMessage);
      }
  }

  const toggleForm = () => {
    setFormSelected(formSelected === "login" ? "sign-up" : "login")
  }

  return (
    <div className="flex flex-auto w-full">
      <Card className={"max-w-full w-[350px]"} radius={"none"}>
        <ErrorMessageBar message={errorMessage} visible={isError} data-testid="error-bar"/>
        <CardBody className="overflow-hidden">
          <Tabs
            fullWidth
            aria-label="Tabs form"
            selectedKey={formSelected}
            size="md"
            onSelectionChange={setFormSelected}
          >
            <Tab key="login" title="Login" data-testid="login-tab">
             <InputBoxes
                    formSelected={formSelected}
                    setUsername={setUsername}
                    setPassword={setPassword}
                    setEmail={setEmail}
                    onSubmitFunc={handleLogin}
                    buttonText="Login"
                    toggleForm={toggleForm}
                    username={username}
                    password={password}
                    email={email}
                />
            </Tab>
            <Tab key="sign-up" title="Sign up" data-testid="signup-tab">
              <InputBoxes
                  formSelected={formSelected}
                  setUsername={setUsername}
                  setPassword={setPassword}
                  setEmail={setEmail}
                  onSubmitFunc={handleSignUp}
                  buttonText="Sign up"
                  toggleForm={toggleForm}
                  username={username}
                  password={password}
                  email={email}
              />
            </Tab>
          </Tabs>
        </CardBody>
      </Card>
    </div>
  );
};

const InputBoxes = ({ formSelected, username, setUsername, password, setPassword, email, setEmail, onSubmitFunc, buttonText, toggleForm }) => {
  return (<form className="flex flex-col gap-4" onSubmit={onSubmitFunc}>
    <Input
        isRequired
        label="Username"
        placeholder="Enter your username"
        type="username"
        data-testid="username-input"
        value={username}
        onChange={(e) => {
          setUsername(e.target.value)
        }}
    />
    {formSelected === "sign-up" ? <Input
        isRequired
        label="Email"
        placeholder="Enter your email"
        type="email"
        value={email}
        data-testid="email-input"
        onChange={(e) => {setEmail(e.target.value)}}
        hidden={formSelected === "login"}
    /> : null}
    <Input
        isRequired
        label="Password"
        placeholder="Enter your password"
        type="password"
        value={password}
        data-testid="password-input"
        onChange={(e) => {
          setPassword(e.target.value)
        }}
    />
    <p className="text-center text-small">
      {buttonText === "Login" ? "Need to create an account? " : "Already have an account? "}
      <Link size="sm" onPress={toggleForm} data-testid="tab-swap-link">
        {buttonText === "Login" ? "Sign up" : "Login"}
      </Link>
    </p>
    <div className="flex gap-2 justify-end">
      <Button fullWidth color="primary" type="submit" data-testid="submit-button">
        {buttonText}
      </Button>
    </div>
  </form>
  );
}
