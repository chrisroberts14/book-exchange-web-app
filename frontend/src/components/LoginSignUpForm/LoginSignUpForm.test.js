import {fireEvent, render, screen, waitFor} from "@testing-library/react";
import {login, signup} from "../../api/api_calls";
import { LoginSignUpForm } from "./LoginSignUpForm.jsx";

jest.mock("../../api/api_calls");

describe("LoginSignUpForm", () => {
  let setUser;

  beforeEach(() => {
    setUser = jest.fn();
    console.log = jest.fn();
    console.error = jest.fn();
  });

  const tryLogin = async () => {
    render(<LoginSignUpForm setUser={setUser} />);
    fireEvent.click(screen.getByTestId("login-tab"));
    fireEvent.change(screen.getByTestId("login-username"), { target: { value: "test_user" } });
    fireEvent.change(screen.getByTestId("login-password"), { target: { value: "password" } });
    fireEvent.click(screen.getByTestId("login-button"));
  }

  const trySignUp = async () => {
    render(<LoginSignUpForm setUser={setUser} />);
    fireEvent.click(screen.getByTestId("signup-tab"));
    fireEvent.change(screen.getByTestId("signup-username"), { target: { value: "test_user" } });
    fireEvent.change(screen.getByTestId("signup-password"), { target: { value: "password" } });
    fireEvent.change(screen.getByTestId("signup-email"), { target: { value: "test@test.com"}});
    fireEvent.click(screen.getByTestId("signup-button"));
  }

  it("renders correctly", () => {
    const { container } = render(<LoginSignUpForm />);
    expect(container).toMatchSnapshot();
  });

  it("clicking on tabs changes form", async () => {
    render(<LoginSignUpForm setUser={setUser} />);
    fireEvent.click(screen.getByTestId("signup-tab"));
    expect(screen.getByTestId("signup-username")).toBeInTheDocument();
    fireEvent.click(screen.getByTestId("login-tab"));
    expect(screen.getByTestId("login-username")).toBeInTheDocument();
  });

  it("clicking on links changes form", async () => {
    render(<LoginSignUpForm setUser={setUser} />);
    fireEvent.click(screen.getByTestId("signup-link"));
    expect(screen.getByTestId("signup-username")).toBeInTheDocument();
    fireEvent.click(screen.getByTestId("login-link"));
    expect(screen.getByTestId("login-username")).toBeInTheDocument();
  });

  it('should login correctly', async () => {
    login.mockResolvedValue({ username: 'testuser', statusCode: 200 });
    await tryLogin();
    await waitFor(() => expect(setUser).toHaveBeenCalled());
  })

  it('should refuse login when using wrong credentials', async () => {
    login.mockResolvedValue({ statusCode: 401 });
    await tryLogin();
    await waitFor(() => expect(console.error).toHaveBeenCalled());
  })

  it('should refuse login when server error', async () => {
    login.mockImplementation(() => {
      throw new Error();
    });
    await tryLogin();
    await waitFor(() => expect(console.error).toHaveBeenCalled());
  })

  it('should signup properly', async () => {
    signup.mockResolvedValue({username: "test", email: "test@test.com", id: "test-id", statusCode: 201});
    await trySignUp();
    await waitFor(() => expect(setUser).toHaveBeenCalled());
  });

  it('should refuse signup if not 201', async () => {
    signup.mockResolvedValue({ statusCode: 401 });
    await trySignUp();
    await waitFor(() => expect(console.error).toHaveBeenCalled());
  });

  it('should refuse signup when server error', async () => {
    signup.mockImplementation(() => {
      throw new Error();
    });
    await trySignUp();
    await waitFor(() => expect(console.error).toHaveBeenCalled());
  });
});
