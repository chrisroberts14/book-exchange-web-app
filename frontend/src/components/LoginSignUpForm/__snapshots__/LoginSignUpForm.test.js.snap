// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginSignUpForm renders correctly 1`] = `
<div>
  <div
    class="flex flex-auto w-full"
  >
    <div
      class="flex flex-col relative overflow-hidden h-auto text-foreground box-border bg-content1 outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 shadow-medium rounded-none transition-transform-background motion-reduce:transition-none max-w-full w-[350px]"
      tabindex="-1"
    >
      <div
        class="relative flex w-full p-3 flex-auto flex-col place-content-inherit align-items-inherit h-auto break-words text-left subpixel-antialiased overflow-hidden"
      >
        <div
          aria-label="Tabs form"
          class="inline-flex w-full"
          data-slot="base"
        >
          <div
            aria-label="Tabs form"
            aria-orientation="horizontal"
            class="flex p-1 h-fit gap-2 items-center flex-nowrap overflow-x-scroll scrollbar-hide bg-default-100 rounded-medium w-full"
            data-slot="tabList"
            id="react-aria-:r0:"
            role="tablist"
          >
            <button
              aria-controls="react-aria-:r0:-tabpanel-login"
              aria-selected="true"
              class="z-0 w-full px-3 py-1 flex group relative justify-center items-center cursor-pointer transition-opacity tap-highlight-transparent data-[disabled=true]:cursor-not-allowed data-[disabled=true]:opacity-30 data-[hover-unselected=true]:opacity-disabled outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 h-8 text-small rounded-small"
              data-key="login"
              data-selected="true"
              data-slot="tab"
              data-testid="login-tab"
              id="react-aria-:r0:-tab-login"
              role="tab"
              tabindex="0"
              type="button"
            >
              <span
                class="absolute z-0 inset-0 rounded-small bg-background dark:bg-default shadow-small"
                data-slot="cursor"
              />
              <div
                class="relative z-10 whitespace-nowrap transition-colors text-default-500 group-data-[selected=true]:text-default-foreground"
                data-slot="tabContent"
              >
                Login
              </div>
            </button>
            <button
              aria-selected="false"
              class="z-0 w-full px-3 py-1 flex group relative justify-center items-center cursor-pointer transition-opacity tap-highlight-transparent data-[disabled=true]:cursor-not-allowed data-[disabled=true]:opacity-30 data-[hover-unselected=true]:opacity-disabled outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 h-8 text-small rounded-small"
              data-key="sign-up"
              data-slot="tab"
              data-testid="signup-tab"
              id="react-aria-:r0:-tab-sign-up"
              role="tab"
              tabindex="-1"
              type="button"
            >
              <div
                class="relative z-10 whitespace-nowrap transition-colors text-default-500 group-data-[selected=true]:text-default-foreground"
                data-slot="tabContent"
              >
                Sign up
              </div>
            </button>
          </div>
        </div>
        <div
          aria-labelledby="react-aria-:r0:-tab-login"
          class="py-3 px-1 data-[inert=true]:hidden outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2"
          data-focus="false"
          data-focus-visible="false"
          data-slot="panel"
          id="react-aria-:r0:-tabpanel-login"
          role="tabpanel"
        >
          <form
            class="flex flex-col gap-4"
          >
            <div
              class="group flex flex-col w-full"
              data-filled="true"
              data-filled-within="true"
              data-has-elements="true"
              data-has-label="true"
              data-required="true"
              data-slot="base"
            >
              <div
                class="relative w-full inline-flex tap-highlight-transparent shadow-sm px-3 bg-default-100 data-[hover=true]:bg-default-200 group-data-[focus=true]:bg-default-100 min-h-10 rounded-medium flex-col items-start justify-center gap-0 transition-background motion-reduce:transition-none !duration-150 outline-none group-data-[focus-visible=true]:z-10 group-data-[focus-visible=true]:ring-2 group-data-[focus-visible=true]:ring-focus group-data-[focus-visible=true]:ring-offset-2 group-data-[focus-visible=true]:ring-offset-background h-14 py-2"
                data-slot="input-wrapper"
                style="cursor: text;"
              >
                <label
                  class="absolute z-10 pointer-events-none origin-top-left rtl:origin-top-right subpixel-antialiased block text-foreground-500 cursor-text after:content-['*'] after:text-danger after:ml-0.5 rtl:after:ml-[unset] rtl:after:mr-0.5 will-change-auto !duration-200 !ease-out motion-reduce:transition-none transition-[transform,color,left,opacity] group-data-[filled-within=true]:text-default-600 group-data-[filled-within=true]:pointer-events-auto group-data-[filled-within=true]:scale-85 text-small group-data-[filled-within=true]:-translate-y-[calc(50%_+_theme(fontSize.small)/2_-_6px)] pe-2 max-w-full text-ellipsis overflow-hidden"
                  data-slot="label"
                  for="react-aria-:r4:"
                  id="react-aria-:r5:"
                >
                  Username
                </label>
                <div
                  class="inline-flex w-full items-center h-full box-border group-data-[has-label=true]:items-end pb-0.5"
                  data-slot="inner-wrapper"
                >
                  <input
                    aria-label="Username"
                    aria-labelledby="react-aria-:r4: react-aria-:r5:"
                    aria-required="true"
                    class="w-full font-normal bg-transparent !outline-none placeholder:text-foreground-500 focus-visible:outline-none data-[has-start-content=true]:ps-1.5 data-[has-end-content=true]:pe-1.5 text-small group-data-[has-value=true]:text-default-foreground"
                    data-slot="input"
                    data-testid="username-input"
                    id="react-aria-:r4:"
                    placeholder="Enter your username"
                    type="username"
                    value=""
                  />
                </div>
              </div>
            </div>
            <div
              class="group flex flex-col w-full"
              data-filled="true"
              data-filled-within="true"
              data-has-elements="true"
              data-has-label="true"
              data-required="true"
              data-slot="base"
            >
              <div
                class="relative w-full inline-flex tap-highlight-transparent shadow-sm px-3 bg-default-100 data-[hover=true]:bg-default-200 group-data-[focus=true]:bg-default-100 min-h-10 rounded-medium flex-col items-start justify-center gap-0 transition-background motion-reduce:transition-none !duration-150 outline-none group-data-[focus-visible=true]:z-10 group-data-[focus-visible=true]:ring-2 group-data-[focus-visible=true]:ring-focus group-data-[focus-visible=true]:ring-offset-2 group-data-[focus-visible=true]:ring-offset-background h-14 py-2"
                data-slot="input-wrapper"
                style="cursor: text;"
              >
                <label
                  class="absolute z-10 pointer-events-none origin-top-left rtl:origin-top-right subpixel-antialiased block text-foreground-500 cursor-text after:content-['*'] after:text-danger after:ml-0.5 rtl:after:ml-[unset] rtl:after:mr-0.5 will-change-auto !duration-200 !ease-out motion-reduce:transition-none transition-[transform,color,left,opacity] group-data-[filled-within=true]:text-default-600 group-data-[filled-within=true]:pointer-events-auto group-data-[filled-within=true]:scale-85 text-small group-data-[filled-within=true]:-translate-y-[calc(50%_+_theme(fontSize.small)/2_-_6px)] pe-2 max-w-full text-ellipsis overflow-hidden"
                  data-slot="label"
                  for="react-aria-:r9:"
                  id="react-aria-:ra:"
                >
                  Password
                </label>
                <div
                  class="inline-flex w-full items-center h-full box-border group-data-[has-label=true]:items-end pb-0.5"
                  data-slot="inner-wrapper"
                >
                  <input
                    aria-label="Password"
                    aria-labelledby="react-aria-:r9: react-aria-:ra:"
                    aria-required="true"
                    class="w-full font-normal bg-transparent !outline-none placeholder:text-foreground-500 focus-visible:outline-none data-[has-start-content=true]:ps-1.5 data-[has-end-content=true]:pe-1.5 text-small group-data-[has-value=true]:text-default-foreground"
                    data-slot="input"
                    data-testid="password-input"
                    id="react-aria-:r9:"
                    placeholder="Enter your password"
                    type="password"
                    value=""
                  />
                </div>
              </div>
            </div>
            <p
              class="text-center text-small"
            >
              Need to create an account?
              <a
                class="relative inline-flex items-center tap-highlight-transparent outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 text-small text-primary no-underline hover:opacity-80 active:opacity-disabled transition-opacity"
                data-testid="tab-swap-link"
                role="link"
                tabindex="0"
              >
                Sign up
              </a>
            </p>
            <div
              class="flex gap-2 justify-end"
            >
              <button
                class="z-0 group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap font-normal subpixel-antialiased overflow-hidden tap-highlight-transparent data-[pressed=true]:scale-[0.97] outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 px-4 min-w-20 h-10 text-small gap-2 rounded-medium w-full [&>svg]:max-w-[theme(spacing.8)] transition-transform-colors-opacity motion-reduce:transition-none bg-primary text-primary-foreground data-[hover=true]:opacity-hover"
                data-testid="submit-button"
                type="submit"
              >
                Login
              </button>
            </div>
          </form>
        </div>
      </div>
    </div>
  </div>
</div>
`;
